src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/ApiDesignGuidelines/ConsiderPassingBaseTypesAsParameters.cs

```diff
@@ -0,0 +1,214 @@
+                }
+                finally
+                {
+                    foreach (var item in typesUsedPerParameter.Values)
+                    {
+                        item.Free(context.CancellationToken);
+                    }
+                    typesUsedPerParameter.Free(context.CancellationToken);
+                }
+
+                // Local functions
+                void ReportIfTypesAreDifferent(IParameterSymbol parameter, ITypeSymbol suggestedBaseTypeSymbol, IEnumerable<IMethodSymbol> overloads)
+                {
+                    if (SymbolEqualityComparer.Default.Equals(suggestedBaseTypeSymbol, parameter.Type))
+                    {
+                        return;
+                    }
+
+                    if (suggestedBaseTypeSymbol is INamedTypeSymbol suggestedNamedType)
+                    {
+                        var parameterIndex = methodSymbol.GetParameterIndex(parameter);
+                        var matchingOverload = methodSymbol.GetMatchingOverload(overloads, parameterIndex, suggestedNamedType, context.CancellationToken);
+
+                        if (matchingOverload != null)
+                        {
+                            return;
+                        }
+                    }
+
+                    // All calls on this parameter are using the same base type which is different
+                    // from the one declared, so we can simply report.
+                    context.ReportDiagnostic(parameter.CreateDiagnostic(Rule, parameter.Name,
+                        parameter.Type.Name, suggestedBaseTypeSymbol.Name));
+                }
+            });
+        }
+
+        private static ITypeSymbol FindDeclaringType(ISymbol symbol)
+        {
+            if (symbol.IsOverride)
+            {
+                var overridenMember = symbol.GetOverriddenMember();
+
+                // It's possible to have GetOverriddenMember() even when IsOverride returns
+                // true so to play defensive we act as it's declared on current type so we
+                // won't suggest any base type.
+                return overridenMember == null
+                    ? symbol.ContainingType
+                    : FindDeclaringType(overridenMember);
+            }
+
+            var interfaceMembers = symbol.GetExplicitOrImplicitInterfaceImplementations();
+            return interfaceMembers.Length switch
+            {
+                // Member is not overridden nor interface implementation so it must be
+                // implemented on the containing type.
+                0 => symbol.ContainingType,
+                1 => interfaceMembers[0].ContainingType,
+                // False negative: more than one interface were found as declaring this
+                // type so let's assume it is defined on the containing type. Ideally,
+                // we would want to collect the type and make multiple offers to users.
+                _ => symbol.ContainingType,
```

Comment:
@mavasani It'd be nice to handle this case but I am not sure how to best handle this. For example if the code uses `.Count` from `List<T>` we have `ICollection<T>.Count`, `IReadonlyCollection<T>.Count` and `ICollection.Count`.